### ISSUE TITLE
Contextual typing does not flow backward through function return types to the function arguments

### ISSUE THREAD

**Issue by drarmstr** [Microsoft/TypeScript#5256]
Consider the following case:

```
class Entry<T> {
    constructor(callback: (v: T) => number) { }
}
interface Widget {
    list: Entry<{ foo: number }>[];
}

var widget: Widget = {
    list: [
        new Entry((v) => v.foo), // Error accessing v.foo
    ]
};
```

We get a compiler error in the callback accessing `v.foo` that `foo` does not exist on type `{}` since `{}` is the default type when it is unable to infer the type.  However, in this case we know from the `Widget` definition that `T` must be typed to `{foo:number}`.  But, it appears TypeScript is not using this as an inference-site to properly determine `T`.

**Comment by RyanCavanaugh** [comment]
More simply:
```ts
class Entry<T> {
    constructor(callback: (v: T) => number) { }
}
var x: Entry<{foo: number}> = new Entry(v => v.foo);
```

**Comment by RyanCavanaugh** [comment]
Let's go even simpler:
```ts
function foo<T>(x: (n: T) => T): T { return x(undefined); }
var x: string = foo(p => p.substr(2));
```
(placeholder since I need to write a longer explanation)

**Comment by drarmstr** [comment]
@mhegazy, Sorry, what is the implication of the `Question` tag?  Is there a follow-up question for me?

**Comment by RyanCavanaugh** [comment]
Sorry, I forgot to write the follow-up I alluded to.

After walking through this with the spec, I think the rule we would need to support this is that a function call (or constructor call) contextually typed by `T` creates an inference site corresponding to the return type of that function or constructor.

That said, it's not clear what runtime semantics this rule corresponds to, or that it wouldn't be an undesirable breaking change.

**Comment by johnnyreilly** [comment]
It would be fantastic if this could be looked into.

**Comment by malibuzios** [comment]
I'm not sure this is the exact same scenario as discussed here but I run into this relatively simple pattern a lot:
```ts
class GenericClass<T, U, V> {}

let x: GenericClass<number, string, boolean>;

// Here the inferred type arguments are <{}, {}, {}>, not <number, string, boolean>:
x = new GenericClass();
```
Since `x` has already been defined with a particular set of generic arguments, these arguments could be implicitly assumed as defaults to the constructor.

I understand there will be cases where that would be very difficult or even impossible:
```ts
let getInstance = () => new GenericClass()
x = getInstance();
```
but these are relatively rare. Most of the time it's just:
```ts
class Example {
	myMap: Map<string, Array<{n: number}>>;
	
	constructor() {
		this.myMap = new Map<string, Array<{n: number}>>(); // <-- Why repeat? it is already known?
	}
}
```

**Comment by RyanCavanaugh** [comment]
Moving discussion to #11152 which has clearer explanation

**Comment by RyanCavanaugh** [comment]
I believe this is fixed by #16072
